staging/src/k8s.io/apimachinery/pkg/util/net/interface.go

```diff
@@ -96,171 +97,176 @@ func isInterfaceUp(intf *net.Interface) bool {
 	return false
 }
 
-//getFinalIP method receives all the IP addrs of a Interface
-//and returns a nil if the address is Loopback, Ipv6, link-local or nil.
-//It returns a valid IPv4 if an Ipv4 address is found in the array.
-func getFinalIP(addrs []net.Addr) (net.IP, error) {
+func isLoopbackOrPointToPoint(intf *net.Interface) bool {
+	if intf.Flags&(net.FlagLoopback|net.FlagPointToPoint) != 0 {
+		return true
+	}
+	return false
+}
+
+// findGlobalIP returns the first IP on interface that is not a loopback or
+// link local address. Can optionally allow IPv6 addresses. Note: ordering
+// of addresses processed is undefined.
+func findGlobalIP(addrs []net.Addr, allowIPv6 bool) (net.IP, error) {
 	if len(addrs) > 0 {
-		for i := range addrs {
-			glog.V(4).Infof("Checking addr  %s.", addrs[i].String())
-			ip, _, err := net.ParseCIDR(addrs[i].String())
+		for _, a := range addrs {
+			glog.V(4).Infof("Checking address %q", a.String())
+			ip, _, err := net.ParseCIDR(a.String())
 			if err != nil {
 				return nil, err
 			}
-			//Only IPv4
-			//TODO : add IPv6 support
-			if ip.To4() != nil {
-				if !ip.IsLoopback() && !ip.IsLinkLocalMulticast() && !ip.IsLinkLocalUnicast() {
-					glog.V(4).Infof("IP found %v", ip)
-					return ip, nil
-				} else {
-					glog.V(4).Infof("Loopback/link-local found %v", ip)
-				}
-			} else {
-				glog.V(4).Infof("%v is not a valid IPv4 address", ip)
+			if ip.To4() == nil && !allowIPv6 {
+				glog.V(4).Infof("Ignoring IPv6 address '%v' for IPv4 only mode", ip)
+				continue
 			}
-
+			// TODO: Do we need to test IsLinkLocalMulticast or IsInterfaceLocalMulticast?
+			// net.Addrs() indicates that it returns *unicast* addresses (and not multicast).
```

Comment:
I must be missing something. IsGlobalUnicast should work for both those addresses, but you do have to negate the result, e.g.:
`if !ip.IsGlobalUnicast() {
    . . . Ignoring loopback/link-local . . .
}`
Here's what I see for IsGlobalUnicast:
`func (ip IP) IsGlobalUnicast() bool {
	return (len(ip) == IPv4len || len(ip) == IPv6len) &&
		!ip.Equal(IPv4bcast) &&
		!ip.IsUnspecified() &&
		!ip.IsLoopback() &&
		!ip.IsMulticast() &&
		!ip.IsLinkLocalUnicast()
}`
and the IsLinkLocalUnicast() should return true for your 169.254.162.166 address:
`func (ip IP) IsLinkLocalUnicast() bool {
	if ip4 := ip.To4(); ip4 != nil {
		return ip4[0] == 169 && ip4[1] == 254
	}
	return len(ip) == IPv6len && ip[0] == 0xfe && ip[1]&0xc0 == 0x80
}`